lib/mutgen.c

```diff
@@ -433,21 +525,31 @@ mutgen_generate(mutgen_t *self, tsk_table_collection_t *tables, int flags)
         right = 0;
         while (right != edge_right) {
             right = GSL_MIN(edge_right, map_position[map_index + 1]);
-            mu = branch_length * (right - left) * map_rate[map_index];
+            site_left = discrete_sites ? ceil(left) : left;
+            site_right = discrete_sites ? ceil(right) : right;
+            mu = branch_length * (site_right - site_left) * map_rate[map_index];
             branch_mutations = gsl_ran_poisson(self->rng, mu);
             for (l = 0; l < branch_mutations; l++) {
                 /* Rejection sample positions until we get one we haven't seen before. */
-                /* TODO add a maximum number of rejections here */
-                do {
```

Comment:
I quite like the do-loop structure here, it makes it clearer that we're not doing this for a fixed number of iterations. 
```c
do {
    position = gsl_ran_flat(self->rng, site_left, site_right);
    if (discrete_sites) {
          position = floor(position);
    }
    search.site->position = position;
￼    avl_node = avl_search(&self->sites, &search);
    num_rejections++;
} while (avl_node == NULL && num_rejections < max_rejections);

```

I'm not sure this is the right structure though as the rejection sampling is quite confusing. Maybe we should push the position choice out to a function to simplify this.

```
ret = mutgen_generate_position(self, site_left, site_right, &avl_node);
if (ret != 0) {
    goto out;
}
```
and we can then do the discrete and continuous cases separately in this function?